### PR TITLE
i3865: Extend reg_resize_to_opsz

### PR DESCRIPTION
**Pull request by johnfxgalea**
Extends reg_resize_to_opsz to support SIMD resizing.

Issue: #3865

### CODE REVIEW

**Comment by johnfxgalea** [comment]
Are ZMM experimental in this case?

**Comment by hgreving2304** [comment]
> Are ZMM experimental in this case?

In general zmms are fully supported, any issue with it would be considered a real bug.

**Comment by johnfxgalea** [comment]
Okay found my mistake. Nothing to do with zmm being experimental.

**Comment by johnfxgalea** [comment]
Someone run arm tests please

**Comment by derekbruening** [comment]
run arm tests

**Comment by derekbruening** [comment]
@AssadHashmi -- @johnfxgalea has commit access.  Do you know why his pull requests aren't automatically running the arm tests?

@johnfxgalea -- have you tried the magic comment "run arm tests"?  Does it trigger the tests for you?

**Comment by johnfxgalea** [comment]
Hey @derekbruening, unfortunately it does not work. 

Hey @hgreving2304, I moved some code away from drx in the PR. As stated on the issue page, I still do not why resizing code has been included in drx. That code made things simpler now that it is in  opnd_shared.c

File: core/arch/opnd_shared.c [hunk 1]
```diff
@@ -2099,12 +2099,75 @@ reg_32_to_opsz(reg_id_t reg, opnd_size_t sz)
...
+    } else if (reg_is_strictly_zmm(simd_reg)) {
+        return simd_reg - DR_REG_START_ZMM + DR_REG_START_XMM;
+    }
+    CLIENT_ASSERT(false, "not a simd register");
+#endif
+    return DR_REG_INVALID;
+}
+
 reg_id_t
 reg_resize_to_opsz(reg_id_t reg, opnd_size_t sz)
```
Comment (hgreving2304) [inline core/arch/opnd_shared.c]: Please add the new description to the docs in opnd.h

Comment (johnfxgalea) [reply core/arch/opnd_shared.c]: Done.

File: core/arch/opnd.h [hunk 2]
```diff
@@ -2418,13 +2418,19 @@ reg_32_to_opsz(reg_id_t reg, opnd_size_t sz);
 
 DR_API
 /**
- * Given a general-purpose register of any size, returns a register in the same
- * class of the given size.  For example, given \p DR_REG_AX or \p DR_REG_RAX
- * and \p OPSZ_1, this routine will return \p DR_REG_AL.
+ * Given a general-purpose or SIMD register of any size, returns a register in the same
```
Comment (hgreving2304) [inline core/arch/opnd.h]: Not true for ARM. Please add a note that ARM is not yet supported.

File: ext/drx/drx.c [hunk 3]
```diff
@@ -1535,43 +1535,19 @@ get_scatter_gather_info(instr_t *instr, scatter_gather_info_t *sg_info)
 static reg_id_t
 simd_reg_to_zmm(reg_id_t simd_reg)
```
Comment (hgreving2304) [inline ext/drx/drx.c]: No need for these functions anymore, please inline the reg_resize_to_ calls below and remove.

File: suite/tests/api/ir_x86.c [hunk 4]
```diff
@@ -1288,6 +1288,64 @@ test_regs(void *dc)
     ASSERT(reg == DR_REG_SP);
     reg = reg_resize_to_opsz(DR_REG_XBP, OPSZ_2);
     ASSERT(reg == DR_REG_BP);
+
+#ifdef X86
```
Comment (hgreving2304) [inline suite/tests/api/ir_x86.c]: Thanks for the test.

File: ext/drx/drx.c (hunk 3 above)
Comment (johnfxgalea) [reply ext/drx/drx.c]: Do you mean remove these functions? What about backward compatibility?

File: core/arch/opnd.h (hunk 2 above)
Comment (johnfxgalea) [reply core/arch/opnd.h]: Done. We should raise another issue for ARM, or avoid closing the current one.

File: ext/drx/drx.c (hunk 3 above)
Comment (hgreving2304) [reply ext/drx/drx.c]: They're static and not exposed.

Comment (johnfxgalea) [reply ext/drx/drx.c]: oh right... missed that

File: core/arch/opnd.h [hunk 5]
```diff
@@ -2418,13 +2418,20 @@ reg_32_to_opsz(reg_id_t reg, opnd_size_t sz);
...
+ * return \p DR_REG_AL. Given \p DR_REG_XMM0 and \p OPSZ-64, it will return \p
+ * DR_REG_ZMM0.
+ *
  * Returns \p DR_REG_NULL when trying to get the 8-bit subregister of \p
  * DR_REG_ESI, \p DR_REG_EDI, \p DR_REG_EBP, or \p DR_REG_ESP in 32-bit mode.
  * For 64-bit versions of this library, if \p sz == OPSZ_8, returns the 64-bit
  * version of \p reg.
+ *
+ * MMX registers are not yet supported.
+ * Moreover, ARM is not yet supported.
```
Comment (hgreving2304) [inline core/arch/opnd.h]: Sorry now it sounds like ARM is not supported at all. Could you spell it out in a way that makes it clear that only ARM SIMD is not supported in the context of this function... Thanks

File: core/arch/opnd.h (hunk 2 above)
Comment (hgreving2304) [reply core/arch/opnd.h]: Please leave the current issue open and change "Fixes" to "Issues:" in the commit message.

Comment (johnfxgalea) [reply core/arch/opnd.h]: Done

File: ext/drx/drx.c (hunk 3 above)
Comment (johnfxgalea) [reply ext/drx/drx.c]: Done

File: core/arch/opnd.h (hunk 5 above)
Comment (johnfxgalea) [reply core/arch/opnd.h]: Done

**Comment by johnfxgalea** [comment]
@derekbruening I have never seen this error reported by Travis. Is it new?

**Comment by hgreving2304** [comment]
run arm tests

**Comment by hgreving2304** [comment]
run arm tests

**Pull request merged**